script/universal/DoubleNestedMultisigBuilder.sol

```diff
@@ -0,0 +1,152 @@
+     * their signature to a facilitator, who will execute the approval transaction for each
+     * signer safe (see step 2).
+     */
+    function sign(address signerSafe, address intermediateSafe) public {
+        address topSafe = _ownerSafe();
+
+        // Snapshot and restore Safe nonce after simulation, otherwise the data logged to sign
+        // would not match the actual data we need to sign, because the simulation
+        // would increment the nonce.
+        uint256 originalTopNonce = _getNonce(topSafe);
+        uint256 originalIntermediateNonce = _getNonce(intermediateSafe);
+        uint256 originalSignerNonce = _getNonce(signerSafe);
+
+        (Vm.AccountAccess[] memory accesses, Simulation.Payload memory simPayload) =
+            _simulateForSigner(intermediateSafe, topSafe, _buildCalls());
+
+        _postSign(accesses, simPayload);
+        _postCheck(accesses, simPayload);
+
+        // Restore the original nonce.
+        vm.store(topSafe, SAFE_NONCE_SLOT, bytes32(originalTopNonce));
+        vm.store(intermediateSafe, SAFE_NONCE_SLOT, bytes32(originalIntermediateNonce));
+        vm.store(signerSafe, SAFE_NONCE_SLOT, bytes32(originalSignerNonce));
+
+        _printDataToSign(signerSafe, _generateIntermediateSafeApprovalCall(intermediateSafe));
+    }
+
+    /**
+     * Step 1.1 (optional)
+     * ======
+     * Verify the signatures generated from step 1 are valid.
+     * This allow transactions to be pre-signed and stored safely before execution.
+     */
+    function verify(address signerSafe, address intermediateSafe, bytes memory signatures) public view {
+        IMulticall3.Call3[] memory calls = _generateIntermediateSafeApprovalCall(intermediateSafe);
+        _checkSignatures(signerSafe, calls, signatures);
+    }
+
+    /**
+     * Step 2
+     * ======
+     * Execute an approval transaction for a signer safe. This method should be called by a facilitator
+     * (non-signer), once for each of the signer safes involved in the nested multisig,
+     * after collecting a threshold of signatures for each signer safe (see step 1).
+     */
+    function approveInit(address signerSafe, address intermediateSafe, bytes memory signatures) public {
+        IMulticall3.Call3[] memory calls = _generateIntermediateSafeApprovalCall(intermediateSafe);
+
+        (Vm.AccountAccess[] memory accesses, Simulation.Payload memory simPayload) =
+            _executeTransaction(signerSafe, calls, signatures, true);
+
+        _postApprove(signerSafe, intermediateSafe);
+    }
+
+    /**
+     * Step 3
+     * ======
+     * Execute an approval transaction for an intermediate safe. This method should be called by a
+     * facilitator (non-signer), for each of the intermediate safes after all of their approval
+     * transactions have been submitted onchain by their signer safes (see step 2).
+     */
+    function runInit(address intermediateSafe) public {
```

Comment:
Out of curiosity, is there a reason why it's named `runInit` and `approveInit` ? If it's not part of some interface, seems like the naming could be improved to better reflect its purpose.